cirq/work/pauli_string_collector.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import random
+from typing import Optional
+
+import numpy as np
+
+from cirq import circuits, study, ops, value
+from cirq.work import collector
+
+
+class PauliStringSampleCollector(collector.SampleCollector[ops.PauliString,
+                                                           float]):
+    """Estimates the energy of a linear combination of Pauli observables."""
+
+    def __init__(self,
+                 circuit: circuits.Circuit,
+                 terms: value.LinearDict[ops.PauliString],
+                 samples_per_job: int = 1000):
+        """
+        Args:
+            circuit: Produces the state to be tested.
+            terms: The pauli product observables to measure. Their sampled
+                expectations will be scaled by their coefficients and their
+                dictionary weights, and then added up to produce the final
+                result.
+            samples_per_job: How many samples to request at a time.
+        """
+        self._circuit = circuit
+        self._samples_per_job = 1000
+        self._terms = value.LinearDict({
+            abs(k): v * k.coefficient for k, v in terms.items()
+        })
+        self._zeros = collections.defaultdict(lambda: 0)
+        self._ones = collections.defaultdict(lambda: 0)
+
+    def next_job(self) -> Optional[collector.CircuitSampleJob[ops.PauliString]]:
+        pauli = random.choice(list(self._terms.keys()))
```

Comment:
does randomness really help here?  it might lead to frustration in debugging and also means that for work sent engine the order will vary, which I don't think is desirable

Reply:
Replaced with just iterating over the terms.